File: src/test/java/br/com/leonardoferreira/jirareport/integration/UpdateBoardIntegrationTest.java
package br.com.leonardoferreira.jirareport.integration;

import br.com.leonardoferreira.jirareport.base.BaseIntegrationTest;
import br.com.leonardoferreira.jirareport.domain.Board;
import br.com.leonardoferreira.jirareport.domain.request.UpdateBoardRequest;
import br.com.leonardoferreira.jirareport.exception.ResourceNotFound;
import br.com.leonardoferreira.jirareport.extension.WireMockExtension;
import br.com.leonardoferreira.jirareport.factory.BoardFactory;
import br.com.leonardoferreira.jirareport.factory.UpdateBoardRequestFactory;
import br.com.leonardoferreira.jirareport.repository.BoardRepository;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import org.apache.http.HttpStatus;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@ExtendWith({SpringExtension.class, WireMockExtension.class})
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class UpdateBoardIntegrationTest extends BaseIntegrationTest {

    @Autowired
    private BoardFactory boardFactory;

    @Autowired
    private UpdateBoardRequestFactory updateBoardRequestFactory;

    @Autowired
    private BoardRepository boardRepository;

    @Test
    public void updateBoard() throws Exception {
        withDefaultUser(() -> boardFactory.create());
        UpdateBoardRequest request = updateBoardRequestFactory.build();

        // @formatter:off
        RestAssured
                .given()
                    .log().all()
                    .header(defaultUserHeader())
                    .contentType(ContentType.JSON)
                    .body(request)
                .when()
                    .put("/boards/1")
                .then()
                    .log().all()
                    .statusCode(HttpStatus.SC_NO_CONTENT);
        // @formatter:on

        Board board = boardRepository.findById(1L)
                .orElseThrow(ResourceNotFound::new);
        Assertions.assertAll("boardContent",
                () -> Assertions.assertEquals(request.getName(), board.getName()),
                () -> Assertions.assertEquals(request.getStartColumn(), board.getStartColumn()),
                () -> Assertions.assertEquals(request.getEndColumn(), board.getEndColumn()),
                () -> Assertions.assertEquals(request.getFluxColumn().size(), board.getFluxColumn().size()),
                () -> Assertions.assertEquals(request.getIgnoreIssueType().size(), board.getIgnoreIssueType().size()),
                () -> Assertions.assertEquals(request.getEpicCF(), board.getEpicCF()),
                () -> Assertions.assertEquals(request.getEstimateCF(), board.getEstimateCF()),
                () -> Assertions.assertEquals(request.getSystemCF(), board.getSystemCF()),
                () -> Assertions.assertEquals(request.getProjectCF(), board.getProjectCF()),
                () -> Assertions.assertEquals(request.getCalcDueDate(), board.getCalcDueDate()),
                () -> Assertions.assertEquals(request.getIgnoreWeekend(), board.getIgnoreWeekend()),
                () -> Assertions.assertEquals(request.getImpedimentColumns().size(), board.getImpedimentColumns().size()),
                () -> Assertions.assertEquals(request.getDynamicFields().size(), board.getDynamicFields().size()));
    }

}

File: application/src/test/kotlin/br/com/jiratorio/integration/board/UpdateBoardIntegrationTest.kt
package br.com.jiratorio.integration.board

import br.com.jiratorio.Authenticator
import br.com.jiratorio.assertion.BoardAssert.Companion.assertThat
import br.com.jiratorio.restassured.specification.notFound
import br.com.jiratorio.junit.testtype.IntegrationTest
import br.com.jiratorio.domain.request.UpdateBoardRequest
import br.com.jiratorio.dsl.restAssured
import br.com.jiratorio.exception.ResourceNotFound
import br.com.jiratorio.factory.domain.entity.BoardFactory
import br.com.jiratorio.factory.domain.request.UpdateBoardRequestFactory
import br.com.jiratorio.repository.BoardRepository
import io.restassured.http.ContentType
import org.apache.http.HttpStatus
import org.junit.jupiter.api.Test

@IntegrationTest
internal class UpdateBoardIntegrationTest(
    private val boardFactory: BoardFactory,
    private val updateBoardRequestFactory: UpdateBoardRequestFactory,
    private val boardRepository: BoardRepository,
    private val authenticator: Authenticator,
) {

    @Test
    @Throws(Exception::class)
    fun `update board`() {
        authenticator.withDefaultUser { boardFactory.create() }
        val request = updateBoardRequestFactory.create()

        restAssured {
            given {
                header(authenticator.defaultUserHeader())
                contentType(ContentType.JSON)
                body(request)
            }
            on {
                put("/boards/1")
            }
            then {
                statusCode(HttpStatus.SC_NO_CONTENT)
            }
        }

        val board = boardRepository.findByIdOrNull(1L)
            ?: throw ResourceNotFound()

        assertThat(board)
            .hasName(request.name)
            .hasStartColumn(request.startColumn)
            .hasEndColumn(request.endColumn)
            .hasFluxColumn(request.fluxColumn?.map { it.toUpperCase() })
            .hasTouchingColumns(request.touchingColumns)
            .hasWaitingColumns(request.waitingColumns)
            .hasIgnoreIssueType(request.ignoreIssueType)
            .hasEpicCF(request.epicCF)
            .hasEstimateCF(request.estimateCF)
            .hasSystemCF(request.systemCF)
            .hasProjectCF(request.projectCF)
            .hasIgnoreWeekend(request.ignoreWeekend)
            .hasImpedimentColumns(request.impedimentColumns?.map { it.toUpperCase() })
    }

    @Test
    fun `update not found board`() {
        restAssured {
            given {
                header(authenticator.defaultUserHeader())
                contentType(ContentType.JSON)
                body(UpdateBoardRequest("name"))
            }
            on {
                put("/boards/999")
            }
            then {
                spec(notFound())
            }
        }
    }
}
